lib/src/main/kotlin/com/swmansion/starknet/data/types/transactions/TransactionReceipt.kt

```diff
@@ -157,6 +167,10 @@ data class PendingTransactionReceipt(
     override val isPending: Boolean = true,
 
     override val type: TransactionReceiptType = TransactionReceiptType.PENDING,
+
+    override val blockHash: Felt? = null,
+    override val blockNumber: Int? = null,
+    override val status: TransactionStatus? = null,
```

Comment:
I don't like the inclusion of these parameters here when according to the RPC spec they will never exist.

Reply:
Then they will be nulls. This way you don't need to check instance of to check the value of status.

Reply:
Having `status` here is probably fine but I'm unsure about values that are always null. This is kind of confusing.

Reply:
Well, in java handling nulls is not as simple as in kotlin so it'll increase boilerplate in PendingTransactions anyway. If cast to correct type it'd be explicit what fields are available. 

```java
if (blockHash != null) {
    // do smth with block hash
}
```

Reply:
Right, with current code they would need to do
```
if (blockHash != null) {
    // do smth with block hash
}
```
with casting in old java they first need to know which class is the right one and then
```
if (receipt instanceof SomeReceipt) {
 SomeReceipt someReceipt = (SomeReceipt) receipt;
 // do smth with receipt
}
```

Also you don't always need to use all fields, sometimes you just need to use some of them and in this case being able to always refer to those fields is very useful. I added it because I had this problem in my tests, previous solution was rather unergonomic.

Reply:
> with casting in old java they first need to know which class is the right one and then
That's exactly why there's a TransactionReceiptType present in all receipts so it's possible to know the type in java without using instanceof

I still cannot understand the benefit of having these fields declared on __base__ transaction type. We know for a fact from the rpc spec that the pending tx's will never have this fields. This change just creates confusion.

Reply:
Also, receipts will still need to be cast anyway since in rpc invoke, declare and deploy receipts are very different object. So the issue is still present in every non unique field, just not block hash and number.

Reply:
>That's exactly why there's a TransactionReceiptType present in all receipts so it's possible to know the type in java without using instanceof
You'll need if for this anyway, so how is it different from instanceof?

>I still cannot understand the benefit of having these fields declared on base transaction type. We know for a fact from the rpc spec that the pending tx's will never have this fields. This change just creates confusion.
Let's say you have a process that goes over your txs and checks their statuses. With nullable field you can just use it right away, without the type cast. At the same time if you want you can still type cast and you'll know that this field is not nullable.
I understand your point, but without nice pattern matching it makes things harder.

Reply:
I can agree that the transaction status being always available may be useful sometimes 

What if we included two custom transaction statuses `PENDING` and `UNKNOWN` and used these as default where necessary instead of optional values. It is much safer to handle these in java than nulls. 

I still don't agree with inclusion of `blockHash` and `blockNumber`. The flow before your changes was better in my opinion

```java
TransactionReceipt receipt = provider.getReceipt(...);
if (receipt.isSuccessful) {
  AcceptedTransactionReceipt acceptedReceipt = (AcceptedTransactionReceipt) receipt;
  // here you have block hash and number available
}
```
Your approach may remove the cast on 3rd line of my snipped but instead you'd have to do `receipt.transactionHash != null` and I fail to understand how that's better than what I wrote.

Reply:
I don't think adding more statuses changes anything. Nulls are only problematic when you receive them when you expect an object. Having them explicitly is ok in my opinion.

AcceptedTransactionReceipt (Processed) still has nullable block{Number, Hash}, so in your example you still need to handle nulls.

Reply:
I think it depends on the RPC specs. In both newest stable spec and current master branch, blockHash and blockNumber are required fields and cannot be null. If we can get confirmation from StarkWare that this is incorrect and will be changed, I can accept this as a temporary solution.

However if they are non-optional on purpose, Gateway receipt shouldn't extend `ProccesedTransactionReceipt` and only there should `blockHash` and `blockNumber` be optional.